libvast/test/index_config.cpp

```diff
@@ -0,0 +1,48 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#define SUITE index_config
+
+#include "vast/index_config.hpp"
+
+#include "vast/concept/convertible/data.hpp"
+#include "vast/data.hpp"
+#include "vast/test/test.hpp"
+
+#include <caf/test/dsl.hpp>
+
+using namespace vast;
+
+namespace {
+
+auto example_index_config = R"__(
+rules:
+  - targets:
+      - suricata.dns.dns.rrname
+      - :address
+    fp-rate: 0.005
+  - targets:
+      - zeek.conn.id.orig_h
+)__";
+
+} // namespace
+
+TEST(example configuration) {
+  const auto yaml = unbox(from_yaml(example_index_config));
+  index_config config;
+  REQUIRE_EQUAL(convert(yaml, config), caf::none);
+  REQUIRE_EQUAL(config.rules.size(), 2u);
+  const auto& rule0 = config.rules[0];
+  REQUIRE_EQUAL(rule0.targets.size(), 2u);
+  CHECK_EQUAL(rule0.targets[0], "suricata.dns.dns.rrname");
+  CHECK_EQUAL(rule0.fp_rate, 0.005);
+  const auto& rule1 = config.rules[1];
+  REQUIRE_EQUAL(rule1.targets.size(), 1u);
+  CHECK_EQUAL(rule1.targets[0], "zeek.conn.id.orig_h");
+  CHECK_EQUAL(rule1.fp_rate, 0.01); // default
```

Comment:
There's a bug in convertible here, which is why the unit tests are failing. Do you mind if I push to this branch directly?

Reply:
Turned out to be a rather simple fix, applying this diff makes all tests pass locally.

```diff
diff --git a/libvast/test/convertible.cpp b/libvast/test/convertible.cpp
index 017c2eb38..9260f625c 100644
--- a/libvast/test/convertible.cpp
+++ b/libvast/test/convertible.cpp
@@ -145,7 +145,7 @@ TEST(failing) {
   x.value.value = 1337;
   r = record{{"value", caf::none}};
   CHECK_EQUAL(convert(r, x), ec::no_error);
-  CHECK_EQUAL(x.value.value, 0);
+  CHECK_EQUAL(x.value.value, 1337);
 }
 
 struct MultiMember {
@@ -354,20 +354,20 @@ struct CafOpt {
 };
 
 TEST(std::optional member variable) {
-  auto x = StdOpt{integer{22}};
+  auto x = StdOpt{integer{42}};
   auto r = record{{"value", caf::none}};
   REQUIRE_EQUAL(convert(r, x), ec::no_error);
-  CHECK_EQUAL(x.value, std::nullopt);
+  CHECK_EQUAL(x.value, integer{42});
   r = record{{"value", integer{22}}};
   REQUIRE_EQUAL(convert(r, x), ec::no_error);
   CHECK_EQUAL(x.value->value, 22);
 }
 
 TEST(caf::optional member variable) {
-  auto x = CafOpt{integer{22}};
+  auto x = CafOpt{integer{42}};
   auto r = record{{"value", caf::none}};
   REQUIRE_EQUAL(convert(r, x), ec::no_error);
-  CHECK_EQUAL(x.value, caf::none);
+  CHECK_EQUAL(x.value, integer{42});
   r = record{{"value", integer{22}}};
   REQUIRE_EQUAL(convert(r, x), ec::no_error);
   CHECK_EQUAL(x.value->value, 22);
@@ -634,9 +634,9 @@ TEST(record with list to optional vector) {
   CHECK(x.xs.contains("baz"));
   CHECK(x.xs["foo"].ovs);
   CHECK_EQUAL(x.xs["foo"].ovs->size(), 3u);
-  CHECK(!x.xs["foo"].ou);
+  CHECK_EQUAL(x.xs["foo"].ou, 0ull);
   CHECK(x.xs["bar"].ovs);
-  CHECK_EQUAL(x.xs["bar"].ou, caf::none);
+  CHECK_EQUAL(x.xs["bar"].ou, 0ull);
   CHECK_EQUAL(x.xs["bar"].ovs->size(), 3u);
   CHECK(!x.xs["baz"].ovs);
   CHECK_EQUAL(*x.xs["baz"].ou, 42u);
diff --git a/libvast/vast/concept/convertible/data.hpp b/libvast/vast/concept/convertible/data.hpp
index 55b2d38bc..164fa45eb 100644
--- a/libvast/vast/concept/convertible/data.hpp
+++ b/libvast/vast/concept/convertible/data.hpp
@@ -437,9 +437,7 @@ public:
   caf::error apply(const record_type::field_view& field, To& dst) {
     auto f = detail::overload{
       [&](const caf::none_t&, const none_type&) -> caf::error {
-        // If the data is nil then the best we can do is default-construct.
-        if constexpr (std::is_default_constructible_v<To>)
-          new (&dst) To{};
+        // If the data is nil then we leave the value untouched.
         return caf::none;
       },
       [&]<class Data>(const Data&, const none_type&) -> caf::error {
@@ -452,9 +450,7 @@ public:
                                            field));
       },
       [&]<concrete_type Type>(const caf::none_t&, const Type&) -> caf::error {
-        // If the data is nil then the best we can do is default-construct.
-        if constexpr (std::is_default_constructible_v<To>)
-          new (&dst) To{};
+        // If the data is nil then we leave the value untouched.
         return caf::none;
       },
       [&]<class Data, concrete_type Type>(const Data& d,
```

Reply:
Thanks a lot for shipping the patch as part of this review!